scripts/gitlab_runner_pre_build/pre_build.py

```diff
@@ -48,6 +48,42 @@ def _token_to_sts_request(raw_jwt, decoded_jwt):
     return assume_role_kwargs
 
 
+def _durable_assume_role_request(assume_role_kwargs):
+    attempts = 0
+    max_attempts = 5
+
+    while attempts < max_attempts:
+        try:
+            req = urllib.request.Request(
+                "https://sts.amazonaws.com/?Action=AssumeRoleWithWebIdentity&Version=2011-06-15&"
+                + urllib.parse.urlencode(assume_role_kwargs),
+                headers={"Accept": "application/json"},
+            )
+
+            with urllib.request.urlopen(req) as response:
+                if response.getcode() == 200:
+                    response = json.loads(response.read().decode("utf-8"))
+                    return response
+                else:
+                    raise ValueError(
+                        f"Unexpected response code from assume role request: {response.getcode()}"
+                    )
+        except urllib.error.HTTPError as e:
+            print(e.read().decode("utf-8"), file=sys.stderr)
+
+            if e.code == 400:
+                attempts += 1
+                print(
+                    f"Assume role request failed with 400, retrying ({attempts}/{max_attempts})",
+                    file=sys.stderr,
+                )
+                time.sleep(2 ** (1 + attempts))
```

Comment:
Assuming I'm reading this correctly, it allows around 2 minutes of "retry grace", assuming all attempts are used.  Is that long enough for the length of outtage we have been seeing?

Reply:
It's tough to say, because a lot of the outages we're seeing aren't 100% down, but very flaky responsiveness. What do you think? Is it worth sleeping for longer to potentially save a job?

Reply:
I'm not sure either.  We already try jobs three times, and once a job fails, there's already some delay before it is retried. Typically I see a failed job go into a paused state, and other jobs that have been queued run first before the next try.  If you added two more attempts using the same backoff computation, that would only be around 8.5 minutes of retrying, total.  What do you think about that?

Reply:
I added one more attempt so we met in the middle 😅.

Reply:
:ok_hand: